static/js/passed-state.js

```diff
@@ -0,0 +1,131 @@
+class PassedState {
+    _key = 'python-type-challenges';
+
+    /**
+     * Initializing when there is no state in the local storage. If there is no state in the local storage, the initial state is required.
+     * this function will check the new state and the old state whether is undefined or not and updated the old state based on new state.
+     * 
+     * @param {object} initialState - the initial state of the challenges which grouped by the level.
+     * @returns void
+     */
+    init(initialState) {
+        const currentState = this.get();
+        // initialize the state when there is no state in the local storage.
+        if (!currentState && !initialState) {
+            throw new Error('initial state is required when there is no state in the local storage.');
+        }
+
+        // add passed property to the challenges in the initial state.
+        const rawState = this._prepareState(initialState);
```

Comment:
The naming of these variables are very confusing.

- `initialState` (in `init`) -> `rawState` (in `_prepareState`) -- returns `state` --> `rawState` (in initialState)
- `rawState` (in `init`) -> `newState` (in `_checkAndMerge`)
- `currentState` (in init) --> `oldState` (in `_checkAndMerge`)

Keeping variable names consistent make code easier to read. Also, removing the `_prepareState` function should help with it too (see my next comment)